### ISSUE TITLE
Oauth - Tokens and codes are created too late?

### ISSUE THREAD

**Issue by joaomatossilva** [aspnet/AspNetKatana#85]
HI,

The Tokens and the authorization codes that are created on each provider, are being called only on the `OnSendingHeaders` event. 

Should this be the correct time to create the codes?

The way I see it, at this time, most of the asp.net request is already finished, and thus a IoC containers are or have already disposed their lifetimescope, making it hard to create providers per request (the only solution here are to create them transient).

Why are the codes being request to be created so late (after others), and not during the normal middleware execution time?

**Comment by Tratcher** [comment]
You're referring to this code path, correct?
https://github.com/aspnet/AspNetKatana/blob/3580ddd66ca91b495d1c1f23f193d7cf5a9a3f42/src/Microsoft.Owin.Security.OAuth/OAuthAuthorizationServerHandler.cs#L198

This is an architectural limitation. Katana auth middleware are invoked lazily either when the app starts writing to the response body (OnSendingHeaders), or when the pipeline unwinds without writing to the body (https://github.com/aspnet/AspNetKatana/blob/3580ddd66ca91b495d1c1f23f193d7cf5a9a3f42/src/Microsoft.Owin.Security/Infrastructure/AuthenticationHandler.cs#L178).

The OAuth server can't run this code any earlier without making significant changes to the shared auth infrastructure.

**Comment by joaomatossilva** [comment]
@Tratcher Yes, that is correct, but also any other provider, like the AccessToken or RefreshToken..

I'm glad that's recognized as a architectural limitation. I was able to workaround it using the https://github.com/autofac/Autofac.Owin/issues/13 discussion here. But it's a pretty big workaround.

The most common scenario on web applictions regading database connections or unit of works, would be on a per request basis, and this architature design is dificulting that. 

I can't say I've tested all IoC frameworks out there to realize if this is an issue of all or just Autofac, but I do beleive it's an issue.

If tokens and codes should be saved, then they should be created on somewhere where it's easy to save them, and not too late on the request lifetime.
